sklearn/preprocessing/_encoders.py

```diff
@@ -89,55 +88,49 @@ def _fit(self, X, handle_unknown='error'):
             else:
                 cats = np.array(self._categories[i], dtype=Xi.dtype)
                 if Xi.dtype != object:
                     if not np.all(np.sort(cats) == cats):
                         raise ValueError("Unsorted categories are not "
                                          "supported for numerical categories")
                 if handle_unknown == 'error':
-                    diff = _encode_check_unknown(Xi, cats)
-                    if diff:
+                    try:
+                        _nanencode(Xi, cats, encode=True,
+                                   missing_values=missing_values)
+                    except ValueError as e:
+                        diff = e.args[0]
                         msg = ("Found unknown categories {0} in column {1}"
                                " during fit".format(diff, i))
                         raise ValueError(msg)
             self.categories_.append(cats)
 
-    def _transform(self, X, handle_unknown='error'):
+    def _transform(self, X, missing_values, handle_unknown='error'):
         X_list, n_samples, n_features = self._check_X(X)
 
         X_int = np.zeros((n_samples, n_features), dtype=np.int)
-        X_mask = np.ones((n_samples, n_features), dtype=np.bool)
+        X_missing_mask = np.zeros((n_samples, n_features), dtype=np.bool)
+        X_unknown_mask = np.zeros((n_samples, n_features), dtype=np.bool)
+        encode_unknown = handle_unknown != 'error'
 
         for i in range(n_features):
             Xi = X_list[i]
-            diff, valid_mask = _encode_check_unknown(Xi, self.categories_[i],
-                                                     return_mask=True)
-
-            if not np.all(valid_mask):
-                if handle_unknown == 'error':
-                    msg = ("Found unknown categories {0} in column {1}"
-                           " during transform".format(diff, i))
-                    raise ValueError(msg)
-                else:
-                    # Set the problematic rows to an acceptable value and
-                    # continue `The rows are marked `X_mask` and will be
-                    # removed later.
-                    X_mask[:, i] = valid_mask
-                    # cast Xi into the largest string type necessary
-                    # to handle different lengths of numpy strings
-                    if (self.categories_[i].dtype.kind in ('U', 'S')
-                            and self.categories_[i].itemsize > Xi.itemsize):
-                        Xi = Xi.astype(self.categories_[i].dtype)
-                    else:
-                        Xi = Xi.copy()
+            encode_results = _nanencode(Xi, self.categories_[i],
+                                        missing_values=missing_values,
+                                        encode=True,
+                                        encode_unknown=encode_unknown)
+
+            if len(encode_results) == 4:
```

Comment:
This is not very readable. (It's also a bit strange that `encode_results` should change the shape of its return value without more explicit request)